File: internal/tools/fleet.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
)

func (r *Registry) registerFleetTools() {
	r.register(ToolDef{
		Name:        "fleet_create",
		Description: "Launch multiple browser sessions in parallel as a fleet",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"count":         map[string]any{"type": "integer", "description": "Number of sessions to launch", "minimum": 1, "maximum": 50},
				"proxy_type":    map[string]any{"type": "string", "enum": []string{"residential", "datacenter", "isp", "mobile"}},
				"proxy_country": map[string]any{"type": "string", "description": "ISO country code"},
				"stealth":       map[string]any{"type": "string", "enum": []string{"none", "standard", "max"}, "default": "max"},
			},
			"required": []string{"count"},
		},
		Handler: r.handleFleetCreate,
	})

	r.register(ToolDef{
		Name:        "fleet_status",
		Description: "Get the status of a fleet and all its sessions",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"fleet_id": map[string]any{"type": "string", "description": "Fleet ID"},
			},
			"required": []string{"fleet_id"},
		},
		Handler: r.handleFleetStatus,
	})

	r.register(ToolDef{
		Name:        "fleet_destroy",
		Description: "Destroy all sessions in a fleet",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"fleet_id": map[string]any{"type": "string", "description": "Fleet ID"},
			},
			"required": []string{"fleet_id"},
		},
		Handler: r.handleFleetDestroy,
	})
}

func (r *Registry) handleFleetCreate(ctx context.Context, args map[string]any) (*ToolResult, error) {
	body := map[string]any{
		"count":   args["count"],
		"stealth": "max",
	}
	if v, ok := args["proxy_type"]; ok {
		proxy := map[string]any{"type": v}
		if country, ok := args["proxy_country"]; ok {
			proxy["country"] = country
		}
		body["proxy"] = proxy
	}
	if v, ok := args["stealth"]; ok {
		body["stealth"] = v
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/fleet", body)
	if err != nil {
		return nil, fmt.Errorf("creating fleet: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleFleetStatus(ctx context.Context, args map[string]any) (*ToolResult, error) {
	fleetID, _ := args["fleet_id"].(string)
	if fleetID == "" {
		return nil, fmt.Errorf("fleet_id is required")
	}

	resp, err := r.client.Do(ctx, "GET", "/v1/fleet/"+fleetID, nil)
	if err != nil {
		return nil, fmt.Errorf("getting fleet status: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleFleetDestroy(ctx context.Context, args map[string]any) (*ToolResult, error) {
	fleetID, _ := args["fleet_id"].(string)
	if fleetID == "" {
		return nil, fmt.Errorf("fleet_id is required")
	}

	resp, err := r.client.Do(ctx, "DELETE", "/v1/fleet/"+fleetID, nil)
	if err != nil {
		return nil, fmt.Errorf("destroying fleet: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

File: src/tools/fleet.ts
import type { Tool } from "@modelcontextprotocol/sdk/types.js";
import type { MeshbrowClient } from "../client.js";

export const fleetTools: Tool[] = [
  {
    name: "fleet_create",
    description:
      "Create a fleet of browser sessions distributed across multiple countries/proxies. Useful for parallel scraping or multi-geo testing.",
    inputSchema: {
      type: "object" as const,
      properties: {
        name: {
          type: "string",
          description: "Fleet name for identification",
        },
        count: {
          type: "number",
          description: "Number of browser sessions to create (default: 5)",
        },
        stealth: {
          type: "string",
          enum: ["none", "basic", "max"],
          description: "Stealth level for all sessions (default: max)",
        },
        proxy_type: {
          type: "string",
          enum: ["residential", "datacenter", "isp", "mobile"],
          description: "Proxy type for all sessions (default: residential)",
        },
        countries: {
          type: "array",
          items: { type: "string" },
          description: "Countries to distribute sessions across (ISO alpha-2 codes)",
        },
        timeout: {
          type: "number",
          description: "Session timeout in minutes (default: 30)",
        },
      },
    },
  },
  {
    name: "fleet_status",
    description: "Get the status of a fleet and all its sessions.",
    inputSchema: {
      type: "object" as const,
      properties: {
        fleet_id: {
          type: "string",
          description: "The fleet ID to check status for",
        },
      },
      required: ["fleet_id"],
    },
  },
  {
    name: "fleet_destroy",
    description: "Destroy a fleet and terminate all its sessions.",
    inputSchema: {
      type: "object" as const,
      properties: {
        fleet_id: {
          type: "string",
          description: "The fleet ID to destroy",
        },
      },
      required: ["fleet_id"],
    },
  },
];

export async function handleFleetTool(
  client: MeshbrowClient,
  name: string,
  params: Record<string, unknown>
): Promise<unknown> {
  switch (name) {
    case "fleet_create": {
      const body: Record<string, unknown> = {
        name: params.name,
        count: params.count || 5,
        config: {
          stealth: params.stealth || "max",
          timeout: params.timeout || 30,
        },
      };

      if (params.proxy_type || params.countries) {
        body.distribution = {
          proxy_types: [params.proxy_type || "residential"],
          countries: params.countries,
        };
      }

      return client.post("/v1/fleet", body);
    }

    case "fleet_status":
      return client.get(`/v1/fleet/${params.fleet_id}`);

    case "fleet_destroy":
      return client.delete(`/v1/fleet/${params.fleet_id}`);

    default:
      throw new Error(`Unknown fleet tool: ${name}`);
  }
}
